File: examples/python-greeting-service/src/greeting_service/domain/greeting.py
"""Dominio de saludo: puro, sin FastAPI. Mismo patrón que los demás
ejemplos del framework, para que sean comparables entre sí.
"""

from __future__ import annotations

import re

_TEMPLATES = {
    "es": lambda name: f"Hola, {name}.",
    "en": lambda name: f"Hello, {name}.",
}
SUPPORTED_LOCALES = tuple(_TEMPLATES.keys())

_MAX_NAME_LENGTH = 80
_VALID_NAME = re.compile(r"^[^\W\d_](?:[^\W\d_]|['\- ])*$", re.UNICODE)


class GreetingError(Exception):
    """Error de negocio: entrada inválida según las reglas del dominio."""


def build_greeting(name: str | None, locale: str = "es") -> str:
    if name is None or not name.strip():
        raise GreetingError("El nombre es obligatorio.")
    trimmed = name.strip()
    if len(trimmed) > _MAX_NAME_LENGTH:
        raise GreetingError(f"El nombre supera {_MAX_NAME_LENGTH} caracteres.")
    if not _VALID_NAME.match(trimmed):
        raise GreetingError("El nombre contiene caracteres no permitidos.")
    effective_locale = (locale or "es").lower()
    template = _TEMPLATES.get(effective_locale)
    if template is None:
        supported = ", ".join(SUPPORTED_LOCALES)
        raise GreetingError(f"Idioma no soportado: se admite {supported}.")
    return template(trimmed)

File: examples/flutter-greeting-app/lib/domain/greeting.dart
/// Dominio de saludo: puro, sin Flutter. Mismo patrón que los demás
/// ejemplos del framework, para que sean comparables entre sí.
library;

class GreetingError implements Exception {
  final String message;
  const GreetingError(this.message);

  @override
  String toString() => message;
}

const List<String> supportedLocales = ['es', 'en'];

const int _maxNameLength = 80;
final RegExp _validName = RegExp(r"^[\p{L}][\p{L}' -]*$", unicode: true);

String buildGreeting(String? name, {String locale = 'es'}) {
  final trimmed = name?.trim() ?? '';
  if (trimmed.isEmpty) {
    throw const GreetingError('El nombre es obligatorio.');
  }
  if (trimmed.length > _maxNameLength) {
    throw GreetingError('El nombre supera $_maxNameLength caracteres.');
  }
  if (!_validName.hasMatch(trimmed)) {
    throw const GreetingError('El nombre contiene caracteres no permitidos.');
  }
  final effectiveLocale = locale.toLowerCase();
  if (!supportedLocales.contains(effectiveLocale)) {
    final supported = supportedLocales.join(', ');
    throw GreetingError('Idioma no soportado: se admite $supported.');
  }
  return switch (effectiveLocale) {
    'en' => 'Hello, $trimmed.',
    _ => 'Hola, $trimmed.',
  };
}
